### PR TITLE
SQL support for custom checks

### PR DESCRIPTION
**Pull request by HEskandari**

### CODE REVIEW

File: src/ServiceControl/CustomChecks/Web/StatisticsResult.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+namespace ServiceControl.CustomChecks
+{
+    using System.Collections.Generic;
+
+    public class StatisticsResult
```
Comment (mikeminutillo) [inline src/ServiceControl/CustomChecks/Web/StatisticsResult.cs]: This pattern might come up a lot. I wonder if this should be something like:

```csharp
class QueryResult<T>
{
  public IList<T> Results { get; set; }
  public int TotalResults { get; set; }
  public string ETag { get; set; }
}
```

We do already have something like this for scatter-gather.

We may want to hold off on doing this if we only have one instance at the moment.

File: src/ServiceControl/CustomChecks/SqlDbCustomCheckDataStore.cs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+                        CustomCheckId = detail.CustomCheckId,
+                        Category = detail.Category,
+                        SucceededAt = detail.ReportedAt,
+                        OriginatingEndpoint = detail.OriginatingEndpoint
+                    }).ConfigureAwait(false);
+                }
+            }
+        }
+
+        public async Task<StatisticsResult> GetStats(HttpRequestMessage request, string status = null)
```
Comment (mikeminutillo) [inline src/ServiceControl/CustomChecks/SqlDbCustomCheckDataStore.cs:98]: Would it be possible to push the `PagingInfo` abstraction to this call (and ideally to the interface)? That would prevent the Http abstraction from pushing down this far.

File: src/ServiceControl/CustomChecks/SqlDbCustomCheckDataStore.cs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        public SqlDbCustomCheckDataStore(string connectionString, IDomainEvents domainEvents, EndpointDetailsMapper mapper)
+        {
+            this.connectionString = connectionString;
+            this.domainEvents = domainEvents;
+            this.mapper = mapper;
+        }
+
+        public async Task UpdateCustomCheckStatus(CustomCheckDetail detail)
+        {
+            var publish = false;
```
Comment (mikeminutillo) [inline src/ServiceControl/CustomChecks/SqlDbCustomCheckDataStore.cs:23]: I wonder if this flag (and the details of the custom check) should be returned to the caller. That way all of the publishing logic can happen outside of the storage.

File: src/ServiceControl/CustomChecks/Web/StatisticsResult.cs (hunk 1 above)
Comment (HEskandari) [reply src/ServiceControl/CustomChecks/Web/StatisticsResult.cs]: I refactored to use the existing QueryResult<T> class.

File: src/ServiceControl/CustomChecks/SqlDbCustomCheckDataStore.cs (hunk 2 above)
Comment (HEskandari) [reply src/ServiceControl/CustomChecks/SqlDbCustomCheckDataStore.cs]: I have changed the interface so it doesn't rely on `HttpRequestMessage` anymore.

File: src/ServiceControl/CustomChecks/SqlDbCustomCheckDataStore.cs (hunk 3 above)
Comment (HEskandari) [reply src/ServiceControl/CustomChecks/SqlDbCustomCheckDataStore.cs]: I have extracted the custom check status notification from the DataStore. It should address this comment, please review the commit [1ff24d8](https://github.com/Particular/ServiceControl/pull/2915/commits/1ff24d8439a8e192583f5bc97755faa809c5a44b).

**Review by mikeminutillo — APPROVED** [review]

**Pull request merged**